### PR TITLE
add ESM support (take 2)

### LINKED ISSUE #1304: Native JS module support in step definitions and support code
### Problem ###
When I attempt to run Cucumber.js with a step definition defined in a native ECMAScript module (as documented [here](https://github.com/cucumber/cucumber-js/blob/master/docs/support_files/step_definitions.md)), the attempt fails with a warning that I am using CommonJS's `require()` on an ES module.

**Cucumber.js version:** 6.0.5
**Node version:** 13.8.0

### Steps to reproduce ###

1. Setup a basic NPM package directory using `npm init` and `npm i cucumber`
    * Set `"type": "module"` in the `package.json` file to ensure that JS files will be treated as native modules
2. Create a basic feature file, `features/mjs.feature`:

        Feature: Native JS Modules
    
            Scenario: Load a native JS module step definition
                Given I have 42 cucumbers in my belly

3. Create a basic step definition file, `features/step_definitions.js`:

        import { Given } from "cucumber";
        
        Given("I have {int} cucumbers in my belly", function (cucumberCount) {
            console.log("Step parsed.");
        });

4. Attempt to run Cucumber:

        $ ./node_modules/.bin/cucumber-js

### Expected result ###

The step definition module should be loaded and used to parse the step.

### Actual result ###

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /some/path/cucumbertest/features/step_definitions.js
require() of ES modules is not supported.
require() of /some/path/cucumbertest/features/step_definitions.js from /some/path/cucumbertest/node_modules/cucumber/lib/cli/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename step_definitions.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /some/path/cucumbertest/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1167:13)
    at Module.load (internal/modules/cjs/loader.js:1000:32)
    at Function.Module._load (internal/modules/cjs/loader.js:899:14)
    at Module.require (internal/modules/cjs/loader.js:1040:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at /some/path/cucumbertest/node_modules/cucumber/lib/cli/index.js:119:42
    at Array.forEach (<anonymous>)
    at Cli.getSupportCodeLibrary (/some/path/cucumbertest/node_modules/cucumber/lib/cli/index.js:119:22)
    at Cli.run (/some/path/cucumbertest/node_modules/cucumber/lib/cli/index.js:141:37)
    at async Object.run [as default] (/some/path/cucumbertest/node_modules/cucumber/lib/cli/run.js:30:14)
```

### Closing ###

For what it's worth, I really appreciate the work that the Cucumber.js team has put in on this project, it has been a major asset to me and my company.  My team has invested some time in building some shared application components in native JS modules, and until recently I had been assuming (based on the `import` syntax in the most current documentation) that this stuff would work with Cucumber when we got around to integrating with our test framework.  That appears not to be the case, though.  Since the shared component is already written and integrated into other parts of our application, I'm reluctant use only step definitions and support code written in CommonJS because of interoperability challenges.

Other things I attempted to get this to work included...

* Giving the native module files `.mjs` extensions, per Node.js convention (this caused them to be overlooked by Cucumber's auto-loading logic, and using `--require` to explicitly load the step definition module threw `Error [ERR_REQUIRE_ESM]: Must use import to load ES Module`
* Wrapping the native modules in CJS modules which use the dynamic `import()` function; the problem was that this approach is asynchronous, and even if the import promise is exported by the CJS module, Cucumber doesn't appear to wait for that promise to resolve before proceeding, because the step was marked as undefined (suggesting that the step definition in the native module was not yet registered).

If it's within my ability, I'd be happy to submit a PR to help resolve this issue, but I'm unfamiliar with the internal workings of Cucumber.js and would appreciate if someone could point me in the right direction.

### LINKED ISSUE #1771: Cannot find cli config file if it has extension .cjs
**Describe the bug**
When configuring the CLI parameters with a config file, this file [must be called `cucumber.js`](https://github.com/cucumber/cucumber-js/blob/main/src/cli/profile_loader.ts#L14)

This is an issue when your Node.js project as [the type `module`](https://nodejs.org/api/packages.html#packages_package_json_and_file_extensions) because it expect JS file to be ESM module rather than CommonJS modules. In such case, `cucumber.js` must be an ESM module, but if so it is not understood and ends up with an error or it must be called `cucumber.cjs` and in such case the cucumber-js cli ignore the file.

**To Reproduce**
1. Set up a new project
2. do `npm init`
3. add the key `type` with the value `module` in the package.json file
4. set up cucumber-js
5. Add a cucumber.js file
6. run `npx cucumber-js`
7. See it fail
8. rename cucumber.js to cucumber.cjs
9. run `npx cucumber-js`
10. See the configuration not being loaded

**Expected behavior**
`cucumber.js` could be alternatively named `cucumber.cjs` or it could be written as a valid ESM module.

### PR DESCRIPTION
**Pull request by davidjgoss**
Support ESM usage with minimal additional flags/configuration. Re-adds the changes originally done in #1589, which were reverted due to the issues described in #1646 (which is now addressed). Fixes #1304.

> To allow importing cucumber functions from ESM support code, provide a "wrapper" mjs entry point for ESM projects as described in https://nodejs.org/api/packages.html#packages_approach_1_use_an_es_module_wrapper.
>
>Usage:
>
>```
>cucumber-js --esm
>```
>
>The `--esm` flag causes us to use `import()` instead of `require()` for loading support code - the former works for both ESM and commonjs, where the latter only works for commonjs.
>
>I think we need to consider usage of this in conjunction with on-the-fly transpilers as out of scope for now. For example ts-node is at a very experimental phase with ESM per https://github.com/TypeStrong/ts-node/issues/1007.
>
>Test project: https://github.com/davidjgoss/cucumber-esm-example

TODO:

- [x] Handle cucumber config file in an ESM project (fixes #1771)
- [x] Test `require-module` compatibility with TypeScript, Babel and CoffeeScript

### CODE REVIEW

**Comment by davidjgoss** [comment]
Added a scenario to prove the issue, where user support code does e.g.:

```js
const TestCaseHookDefinition = require('@cucumber/cucumber/lib/models/test_case_hook_definition')
```

And we get an error like:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/models/test_case_hook_definition' is not defined by "exports" in /Users/davidgoss/Projects/cucumber-js/package.json imported from /Users/davidgoss/Projects/cucumber-js/tmp/direct_imports.feature_49/features/step_definitions/cucumber_steps.js
           at throwExportsNotFound (internal/modules/esm/resolve.js:285:9)
           at packageExportsResolve (internal/modules/esm/resolve.js:508:3)
           at trySelf (internal/modules/cjs/loader.js:400:34)
           at Function.Module._resolveFilename (internal/modules/cjs/loader.js:793:24)
           at Function.Module._load (internal/modules/cjs/loader.js:667:27)
           at Module.require (internal/modules/cjs/loader.js:887:19)
           at require (internal/modules/cjs/helpers.js:74:18)
           at Object.<anonymous> (/Users/davidgoss/Projects/cucumber-js/tmp/direct_imports.feature_49/features/step_definitions/cucumber_steps.js:2:32)
           at Module._compile (internal/modules/cjs/loader.js:999:30)
           at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
           at Module.load (internal/modules/cjs/loader.js:863:32)
           at Function.Module._load (internal/modules/cjs/loader.js:708:14)
           at Module.require (internal/modules/cjs/loader.js:887:19)
           at require (internal/modules/cjs/helpers.js:74:18)
           at World.<anonymous> (/Users/davidgoss/Projects/cucumber-js/features/support/hooks.ts:119:11)
```

This happens whether or not you're using ESM, and is because of the new block added to our `package.json`:

```json
"exports": {
  "import": "./lib/wrapper.mjs",
  "require": "./lib/index.js"
},
```

We need this in order for ESM to work (and for CommonJS to still work), but it does have the effect of enforcing that `require`s can only be done at the entry point, meaning "deep" ones to individual files will fail.

Node has the concept of [subpath patterns](https://nodejs.org/docs/latest-v14.x/api/packages.html#packages_subpath_patterns) for this kind of thing, but this can't be mix-and-matched with the usage of `exports` for ESM vs CommonJS entry points like we have, so that's off the table.

- - -

My current thinking: the public API for cucumber-js is what we expose on the main entry point. Once you `require` other things via their file path, you're into the weeds; package authors understand this. So in principle running with this as it is wouldn't be a breaking change...but it would be disruptive to end users and it's not their fault.

We could instead:

1. Add things to our main entry point to address the needs of package authors (see list in #1646)
2. Submit PRs to those projects to update their imports
3. Release this change again as semver minor

WDYT @charlierudolph @aurelien-reeves ?

**Comment by jan-molak** [comment]
@davidjgoss to avoid "polluting" the main entry point with functions and classes only package authors need to know about you might want to consider exporting them under some object like `internal`. 

So a regular Cucumber user would only see one additional thing exported by the module, for example:

```typescript
import { Given, When, Then, internal } from '@cucumber/cucumber'

const definition = new internal.TestCaseHookDefinition(...)
```

This way you could keep attaching "internal" functions useful for package maintainers without making it disruptive.

Could this work?

**Comment by davidjgoss** [comment]
@jan-molak I think that's a good point yeah. It's still part of the API but at least grouped where it's less commonly used. I think there is some prior art here with `formatterHelpers`

**Comment by TheLudd** [comment]
Have you looked into subpath exports? https://nodejs.org/api/packages.html#packages_subpath_exports
Not sure if it is possible to combine with the import/require statements here https://github.com/cucumber/cucumber-js/pull/1649/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R167 but I think that is the way to expose other exported paths than the main one going forward.

**Comment by davidjgoss** [comment]
Thanks for the pointer @TheLudd - a combination of subpath exports and [export patterns](https://github.com/nodejs/node/pull/34718) gets it done:

```json
{
  "exports": {
    ".": {
      "import": "./lib/wrapper.mjs",
      "require": "./lib/index.js"
    },
    "./lib/*": {
      "require": "./lib/*.js"
    }
  },
}
```

**Comment by davidjgoss** [comment]
@charlierudolph @aurelien-reeves with the fix for deep imports in commonjs this is good to go now.

File: src/importers.js
Comment (aurelien-reeves) [inline src/importers.js:5]: We dropped Node 10 support, so, is the following still required?

**Review by aurelien-reeves — COMMENTED** [review]
This is really great, well done! 👍 

We may have to add an entry in the changelog for bringing back esm support

And, the esm.md is a good idea 👍 
Could we add some code examples in it?

File: src/importers.js
Comment (davidjgoss) [reply src/importers.js:5]: Ah good point, I need to update this comment.

We still need to support the commonjs path unfortunately because of transpilers that people use via `--require-module` - for example with TypeScript, ts-node has [only early experimental support](https://github.com/TypeStrong/ts-node/issues/1007) for handling `import()` (this is why the features in cucumber-js itself didn't work when I tried this).

So I think it's safest to still default to `require()` for now and have `import()` be opt-in via the `--esm` flag.

Comment (davidjgoss) [reply src/importers.js]: Alternatively, we could use `import()` for loading all user code _except_ for requiring support files where `--require-module` is set (meaning `ts-node` or `babel` is in play and thus we'll fall back to `require`).

I think that's better actually, I'll run it up and see what happens.

**Comment by davidjgoss** [comment]
> And, the esm.md is a good idea 👍 Could we add some code examples in it

Good point, I've added example code now.

**Comment by davidjgoss** [comment]
@charlierudolph @aurelien-reeves 

So this is now simplified a bit with Node 10 support having been dropped. `await import()` is now always available so we can avoid the `--esm` flag since it works with both ESM and CommonJS code. A couple of exceptions:

- The `cucumber-js` profiles file remains CommonJS only and loaded via `require()`. Switching it to use `await import()` threw up some issues which I think are related to `default` being one of the profile keys. At any rate this is likely to become a legacy thing via #751 / #1711 so I think it's best to just leave it alone at this point.
- You can't `await import()` things that aren't JavaScript, meaning it would break for folks using TypeScript or Babel. I've used the presence of `--require-module` as a heuristic for this, so if that option has been used (meaning `ts-node` or `babel` etc is in play) then we'll fall back to using `require()` for loading the support code. Can you see any issues with this?

**Comment by aurelien-reeves** [comment]
I'll take a look next week :)

File: src/formatter/builder.ts
Comment (aurelien-reeves) [inline src/formatter/builder.ts:109]:
```suggestion
      const syntaxImporter = await importer(pathToFileURL(fullSyntaxPath))
      Syntax = FormatterBuilder.resolveConstructor(syntaxImporter)
```
Or something like that?
Seeing two times `Syntax = ` puzzled me actually

**Review by aurelien-reeves — COMMENTED** [review]
Looks good!
I have a few comments, but it looks really good.

I see a lot of new usage of `async` `await`. Beside the compatibility with `esm`, may this impact the code of cucumber users and third-party developers?

File: features/esm.feature
Comment (davidjgoss) [reply features/esm.feature:3]: That's right, thanks!

**Comment by davidjgoss** [comment]
> I see a lot of new usage of async await. Beside the compatibility with esm, may this impact the code of cucumber users and third-party developers?

Async/await has been natively supported in Node since v7 so we're all good.

**Comment by aurelien-reeves** [comment]
> 
> 
> > I see a lot of new usage of async await. Beside the compatibility with esm, may this impact the code of cucumber users and third-party developers?
> 
> Async/await has been natively supported in Node since v7 so we're all good.

Yes, I don't have concerns with using async/await.
I just wanted to make sure there won't be breaking changes for existing code of our users and/or third-party developers.

File: docs/esm.md
Comment (charlierudolph) [inline docs/esm.md:30]: Does this have the caveat that you everything must be javascript in order to use ESM? Given you brought it up. Seems like we should document it.

Comment (davidjgoss) [reply docs/esm.md:30]: Good point - I have added a note for this.

File: src/formatter/builder.ts
Comment (davidjgoss) [reply src/formatter/builder.ts]: Agreed - fixed by inlining the import

**Comment by davidjgoss** [comment]
@charlierudolph @aurelien-reeves feedback addressed, and now using the file extension as a more deterministic way to know whether to fall back to `require` - basically if not a JS file, don't try and use native import. Let me know if you think we're good to go.

File: src/cli/index.ts
Comment (aurelien-reeves) [inline src/cli/index.ts:172]: We've lost the `try ... catch` around `require(codePath)` in the process. Is it on purpose?

**Review by aurelien-reeves — APPROVED** [review]
There is a lot of changes, and those are not simple one.
As long as the tests are passing, and there is not a big thing that we would certainly have noticed already, I give it a go despite my small small question

File: src/formatter/builder.ts
Comment (charlierudolph) [inline src/formatter/builder.ts:109]: Why is importer used here always instead of a switch between importer / require? Same for other instances

Comment (davidjgoss) [reply src/formatter/builder.ts:109]: For custom snippet syntaxes and formatters, there should be no need to fall back to `require` - we never supported transpilation for these things, so whether they are ESM or CommonJS modules, the `import` will just work. Support code is only different because of the transpilation.

File: src/cli/index.ts
Comment (davidjgoss) [reply src/cli/index.ts:172]: No - thanks, good catch. Re-added.

File: features/support/hooks.ts
Comment (charlierudolph) [inline features/support/hooks.ts:51]: Why the dedicated return statement?

File: src/runtime/parallel/worker.ts
Comment (charlierudolph) [inline src/runtime/parallel/worker.ts:24]: What makes this need to be a require instead of an import?

**Comment by charlierudolph** [comment]
Looks great! Just a couple comments

File: features/support/hooks.ts
Comment (davidjgoss) [reply features/support/hooks.ts]: I have no idea...removed :)

File: src/runtime/parallel/worker.ts
Comment (petermorlion) [reply src/runtime/parallel/worker.ts:24]: If this is the only comment holding this back from being merged and released, can I help in any way? I could really use ESM support in cucumber in a new project. It would allow me to write the code in ESM instead of CommonJS modules. So if I can help... Do you need someone to try it out with this line changed to use `import`?

Comment (davidjgoss) [reply src/runtime/parallel/worker.ts]: Resolved by allowing JS in the TypeScript config, and adding JSDoc-flavoured type definitions to the file.

**Comment by aurelien-reeves** [comment]
The errors with the Windows builds seems legit :(

**Comment by davidjgoss** [comment]
@aurelien-reeves yeah I think adding `allowJs` might have done this. I'll probably just back that out and we'll live with doing an ugly `require` for now

**Comment by aurelien-reeves** [comment]
🎉

**Comment by nemanja-tosic** [comment]
When will this land? I see it's part of 8.0.0 but there's no date anywhere I've seen.

**Comment by aurelien-reeves** [comment]
You don't need the --esm flag anymore

https://github.com/cucumber/cucumber-js/blob/main/docs/esm.md

**Comment by Ploppz** [comment]
Hmm.. but if I add this package.json file to my support files and steps files directories:

```
{
  "name": "test1",
  "version": "0.0.1",
  "type": "module"
}
```

Then I get this error:
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/node/tests/features/step_definitions/overview_steps.js from /home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/api/support.js not supported.
Instead change the require of overview_steps.js in /home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/api/support.js to a dynamic import() which is available in all CommonJS modules.
    at /home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/api/support.js:14:32
    at Array.map (<anonymous>)
    at getSupportCodeLibrary (/home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/api/support.js:14:18)
    at runCucumber (/home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/api/run_cucumber.js:33:53)
    at async Cli.run (/home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/cli/index.js:45:29)
    at async Object.run [as default] (/home/node/node_modules/.pnpm/@cucumber+cucumber@8.5.1/node_modules/@cucumber/cucumber/lib/cli/run.js:34:18)
```

**Comment by davidjgoss** [comment]
> The default behaviour and the require option both use the [legacy CommonJS modules API](https://nodejs.org/api/modules.html) to load your files. If your files are native ES modules, you'll need to use the import option instead in the same way, and they'll be loaded with the [new ES modules API](https://nodejs.org/api/esm.html). See [ES Modules](https://github.com/cucumber/cucumber-js/blob/main/docs/esm.md) for more on using Cucumber in an ESM project.

**Comment by Ploppz** [comment]
Is that quoted from some source? Sorry if I misunderstand but I do already use `import` when importing anything cucumber-related; does that quote suggest anything else?
My error seems to indicate that cucumber tries to `require` my js files instead of `import`ing them, if I understand correctly.

**Comment by aurelien-reeves** [comment]
I guess the quotation marks from David is a typo.
However the documentation here should help: https://github.com/cucumber/cucumber-js/blob/main/docs/esm.md

You should not use a dedicated package.json file inside your support code directory.

What are you trying to achieve exactly? Your project is using ESM or not?

If not, I suggest not to use it with your support code either.

If it does, you do not need a package.json beside your support code.

The `require` and `import` things are used when you explicitly require or import your support code using configuration file or command-line arguments.

Would you have a repo where we could take a look?

**Comment by Ploppz** [comment]
I was initially confused about `you'll need to use the import configuration option to specify your files, rather than the require option` especially since the below example does not even use said import configuration option, but looking at the documentation for the config cleared that up. I tried the config:
```js
const common = {
  import: ['features/support/**/*.js']
}

export default {
  ...common,
}

export const ci = {
  ...common,
}
```
but I get 
```
export default {
^^^^^^

SyntaxError: Unexpected token 'export'
```
anyway I guess I shouldn't need such a config file?
My project is generally ESM I guess yes, but I say that solely because it consistently uses `import` instead of `require`, I don't know in what else determines whether it's an ESM project.

Well, if I remove the `package.json` files, I get errors:
```
import { When, Then, Given } from '@cucumber/cucumber';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
So I left them in the attached zip file. it's only the contents of the `tests/` folder in my project. I start tests with `./node_modules/.bin/cucumber-js tests`. Do you need anything else?
[tests.zip](https://github.com/cucumber/cucumber-js/files/9363357/tests.zip)

**Comment by aurelien-reeves** [comment]
What is your version of @cucumber/cucumber?
Which version of node-js?

Do you have `"type": "module"` in your root package.json?

Here's the documentation for the configuration: https://github.com/cucumber/cucumber-js/blob/main/docs/configuration.md

The error related to the config file as shown below seems to show that your project is not a module. If you still want to use ESM with cucumber, try renaming your configuration file into `cucumber.mjs`
```
export default {
^^^^^^

SyntaxError: Unexpected token 'export'
```

Please not also that per default, your cucumber configuration file should be at the root level of your project, not in your "tests" folder.

what is your cucumber command and arguments to run your tests? Are you using a npm script?

**Comment by davidjgoss** [comment]
Yeah further to what @aurelien-reeves said, a CRA project's files may use import and export syntax but they aren't really ES modules, since they're compiled down to CommonJS and then bundled for the browser. As such you probably shouldn't add `type: module` to your package.json. Having your Cucumber config and support code files as `.mjs` is probably the way to go here - you're marking those files as ESM without trying to reclassify the whole project.

**Comment by Ploppz** [comment]
Thank you both so much! Alright I changed everything cucumber-related to `.mjs`.
> What is your version of @cucumber/cucumber?
> Which version of node-js?
> what is your cucumber command and arguments to run your tests? Are you using a npm script?

```
// ...
    "@cucumber/cucumber": "8.5.1",
// ...
  "scripts": {
    "test": "./node_modules/.bin/cucumber-js tests",
// ...

  "engines": {
    "node": "17.x"
  }
```

Now I'm getting a new error that seems to stem from inside `node`, with no information about code location in my own code in any case:
```
SyntaxError: Unexpected token ')'
    at ESMLoader.moduleStrategy (node:internal/modules/esm/translators:117:18)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:361:14)
    at async link (node:internal/modules/esm/module_job:70:21)
```
idk if there's still some mismatch with like, node version (since JS seems to be very finicky about all its moving parts)?

**Comment by Ploppz** [comment]
Sorry, the extra ')' was in my own code! I just thought it wouldn't be because it didn't point to any of my files

**Comment by devexpert7** [comment]
Is this fix available in latest release?  Please share working code if available. 
I get same error 
npx cucumber-js
Error: Cucumber expected a CommonJS module at '/Users/dev/features/support/support.js' but found an ES module.
      Either change the file to CommonJS syntax or use the --import directive instead of --require.
      
      Original error message: require() of ES Module /Users/dev/features/support/support.js from /Users/develop/node_modules/@cucumber/cucumber/lib/try_require.js not supported.
Instead change the require of support.js in /Users/dev/gitlab1/node_modules/@cucumber/cucumber/lib/try_require.js to a dynamic import() which is available in all CommonJS modules.
    at tryRequire (/Users/dev/gitlab1/node_modules/@cucumber/cucumber/lib/try_require.js:15:19)

**Comment by unional** [comment]
Hi, @devexpert7,

Probably need more information about your project. Do you have `type: module` in your package json?

**Comment by devexpert7** [comment]
> Hi, @devexpert7,
> 
> Probably need more information about your project. Do you have `type: module` in your package json?

Yes @unional ,   I have `type: module` in your package json. 
I tried to fix this by adding configuration file cucumber.cjs  >
const common = {
    requireModule: [
        ...(
            process.env.CI ? [] : ["chromedriver"]
        )
    ],
    import: [
        "src/steps/*.ts",
        "src/steps",
    ]
}

module.exports = common
....

 getting below error now: 
> my-dev@1.0.0 test-e2e-scenario
> npm run cucumber -- features --name sayHello


> my-dev@1.0.0 cucumber
> ts-node node-modules/@cucumber/cucumber/bin/cucumber.js -p default features --name sayHello

node:internal/modules/cjs/loader:1080
  throw err;
  ^

Error: Cannot find module './cucumber.js'
Require stack:
- /Users/project1/node-modules/@cucumber/cucumber/bin/imaginaryUncacheableRequireResolveScript
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at Function.resolve (node:internal/modules/cjs/helpers:127:19)
    at requireResolveNonCached (/Users/project1/node_modules/ts-node/dist/bin.js:549:16)
    at getProjectSearchDir (/Users/project1/node_modules/ts-node/dist/bin.js:519:40)
    at phase3 (/Users/project1/node_modules/ts-node/dist/bin.js:267:27)
    at bootstrap (/Users/project1/node_modules/ts-node/dist/bin.js:47:30)
    at main (/Users/project1/node_modules/ts-node/dist/bin.js:33:12)
    at Object.<anonymous> (/Users/project1/node_modules/ts-node/dist/bin.js:579:5)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/project1/node-modules/@cucumber/cucumber/bin/imaginaryUncacheableRequireResolveScript'
  ]
}

**Pull request merged**
